crates/cadenza-eval/src/tests.rs

```diff
@@ -0,0 +1,205 @@
+}
+
+/// Helper to evaluate a single expression.
+fn eval_one(src: &str) -> Result<Value, Error> {
+    eval_all(src)?
+        .into_iter()
+        .next()
+        .ok_or_else(|| Error::syntax("no expressions"))
+}
+
+#[test]
+fn test_basic_arithmetic() {
+    assert_eq!(eval_one("1 + 2").unwrap(), Value::Integer(3));
+    assert_eq!(eval_one("10 - 4").unwrap(), Value::Integer(6));
+    assert_eq!(eval_one("3 * 7").unwrap(), Value::Integer(21));
+    assert_eq!(eval_one("20 / 4").unwrap(), Value::Integer(5));
+}
+
+#[test]
+fn test_operator_precedence() {
+    // * has higher precedence than +
+    assert_eq!(eval_one("2 + 3 * 4").unwrap(), Value::Integer(14));
+    // Left-to-right for same precedence
+    assert_eq!(eval_one("10 - 5 - 2").unwrap(), Value::Integer(3));
+}
+
+#[test]
+fn test_float_arithmetic() {
+    assert_eq!(eval_one("1.5 + 2.5").unwrap(), Value::Float(4.0));
+    assert_eq!(eval_one("3.0 * 2.0").unwrap(), Value::Float(6.0));
+}
+
+#[test]
+fn test_mixed_arithmetic() {
+    // Integer + Float = Float
+    assert_eq!(eval_one("1 + 2.5").unwrap(), Value::Float(3.5));
+    assert_eq!(eval_one("2.5 + 1").unwrap(), Value::Float(3.5));
+}
+
+#[test]
+fn test_comparison_operators() {
+    assert_eq!(eval_one("1 < 2").unwrap(), Value::Bool(true));
+    assert_eq!(eval_one("2 < 1").unwrap(), Value::Bool(false));
+    assert_eq!(eval_one("1 <= 1").unwrap(), Value::Bool(true));
+    assert_eq!(eval_one("2 > 1").unwrap(), Value::Bool(true));
+    assert_eq!(eval_one("1 >= 1").unwrap(), Value::Bool(true));
+    assert_eq!(eval_one("1 == 1").unwrap(), Value::Bool(true));
+    assert_eq!(eval_one("1 != 2").unwrap(), Value::Bool(true));
+}
+
+#[test]
+fn test_string_literal() {
+    assert_eq!(
+        eval_one("\"hello world\"").unwrap(),
+        Value::String("hello world".to_string())
+    );
+}
+
+#[test]
+fn test_multiple_expressions() {
+    let results = eval_all("1\n2\n3").unwrap();
+    assert_eq!(results.len(), 3);
+    assert_eq!(results[0], Value::Integer(1));
+    assert_eq!(results[1], Value::Integer(2));
+    assert_eq!(results[2], Value::Integer(3));
+}
+
+#[test]
+fn test_undefined_variable_error() {
+    let result = eval_one("undefined_variable");
+    assert!(result.is_err());
+    match result.unwrap_err() {
+        Error::UndefinedVariable(name) => assert_eq!(name, "undefined_variable"),
+        e => panic!("unexpected error: {e}"),
+    }
+}
+
+#[test]
+fn test_builtin_function() {
+    // Note: "add 1 2" is parsed as ((add 1) 2), i.e., nested applies.
+    // For simple testing, we use a single-argument function.
+    let parsed = parse("inc 5");
+    let root = parsed.ast();
+    let mut interner = Interner::new();
+    let mut env = Env::new();
+    let mut compiler = Compiler::new();
+
+    // Define a builtin inc function (increment by 1)
+    let inc_id = interner.intern("inc");
+    env.define(
+        inc_id,
+        Value::BuiltinFn(BuiltinFn {
+            name: "inc",
+            func: |args| {
```

Comment:
It would be nice if we had a helper to make this a little more native to rust. So you could just do:

```rust
bultin!(fn inc(a: Integer) {
    a + 1
})
```

And that would return a builtin value